updated the ldap bind functions to provide for anonymous binds

babagreensheep opened the pull request:
Addresses [this request for anonymous binds](https://github.com/ViViDboarder/vaultwarden_ldap/issues/81#issuecomment-1132927133)

ViViDboarder commented:
Thanks for the PR! Do you think you could build an integration test case? I could probably help out with it later too.

babagreensheep commented:
Sorry, I'm not quite sure how to do that? This is my first time contributing to open source and I'd love any guidance!

ViViDboarder commented:
Just made some changes to make it easier. If you're more familiar than I am with ldap, you can add an ldif file to the `itest/ldif` directory. That ldiff will need to give anon bind read access. I tried this but couldn't get slapd to start, but I'm not very familiar with ldap.

Once that's working, you'll need to create a new configuration to attempt anon-binding. The easiest way would be to copy `itest/docker-compose.itest-env.yml` and call it something like `docker-compose.itest-anon.yml` and remove the bind username and passwords. Then copy the `itest-env` target in the `Makefile` calling it something like `itest-anon` and change the compose file to the new one.

From there you can do `make itest-anon` and you should be able to debug from there.

ViViDboarder commented on src/main.rs:
```diff
@@ -84,8 +84,11 @@ fn ldap_client(
-    ldap.simple_bind(bind_dn.as_str(), bind_pw.as_str())
-        .context("Could not bind to LDAP server")?;
+
+    if bind_dn.is_some() && bind_pw.is_some() {
+        ldap.simple_bind(&bind_dn.unwrap(), &bind_pw.unwrap())
+            .context("Could not bind to LDAP server")?;
+    }
```
Rather than using `is_some()` and `unwrap()`, we should use `match` here. This is not only more idiomatic, but it allows us to exhaust the possible branches and provide more descriptive errors.

Here's a simple example of what I mean:

```rust
match (bind_dn, bind_pw) {
        (Some(dn), Some(pw)) => println!("bind as {}:{}. Use ldap.simple_bind(...)", dn, pw),
        
        (None, None) => println!("anon bind. Use ldap.bind()"),
        // These last two should raise errors 
        (Some(dn), None) => println!("bind as {} with no pw. Should error.", dn),
        (None, Some(_)) => println!("Invalid! Can't bind with pw only. Should error.")
    }
 ```
Example on playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c12ddb12f80b77d438cc5b043f9e1779

Also, you'll see in the example I note that you should still bind if it's anon. You need to do `ldap.bind()` instead. https://ldap3.readthedocs.io/en/latest/bind.html#anonymous-bind

ViViDboarder reviewed (changes requested):
Thanks for your contribution and congrats on your first open source experience! I've provided some feedback that should fix a bug as well as make it easier to maintain. If you have any questions, I'm happy to help out.

babagreensheep commented:
Okay got it let me work on the test (and fix non-idiomatic code: yes I was lazy earlier). 

Re anon bind: in all my tests (not just with this library but other services I'm working on) I didn't necessarily have to bind to empty strings and in fact on the Rust library I ran into some issues binding with empty strings. I will try to see if I can bottom this out. 

Thanks for helping a first timer out!!

babagreensheep commented:
Hi! I've added a new test (itest-anon) in the makefile and the tests appears to run successfully. I've also updated fixed the non-idiomatic pattern and added errors for invalid authentication (username but no pw and vice versa). I'm not sure if i have correctly pushed it though?

ViViDboarder commented:
Thanks for your contribution! I ran into itest issues relating to the ldap image that I was using, so I switched images and then had to rebase your change on top as well. The new image allows anon by default, simplifying things a lot too.
